### PR TITLE
Add key schedule injection structure

### PR DESCRIPTION
**Pull request by chris-wood**

### CODE REVIEW

File: draft-jhoyla-tls-key-injection.md [hunk 1]
```diff
@@ -53,27 +53,25 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
...
-For example if two parties have a shared PSK, but have not agreed a hash
-algorithm then this process can be used to ensure agreement on the hash
-algorithm before the TLS handshake completes, and further to ensure that both
-parties agree on the selected hash algorithm.
-
-The change to the binder key computation is necessary because otherwise one
-party may believe it has agreed on some authenticated data, but the other
-party may believe that it is simply using a vanilla PSK, and be unaware of the
-extra context implied.
+To achieve this, pre-shared keys modify the binder key computation. This is
```
Comment (jhoyla) [inline draft-jhoyla-tls-key-injection.md:70]: I'm not sure this is quite sharp enough. A key aspect of the change to the binder key computation is the modification of the labels.

File: draft-jhoyla-tls-key-injection.md [hunk 2]
```diff
@@ -53,27 +53,25 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
...
+the TLS 1.3 key schedule.
+
+## Early Secret Injection
 
 TLS provides exporter keys that allow for other protocols to provide
 data authenticated by the TLS channel. This can be used to bind a protocol to a
 specific TLS handshake, giving joint authentication guarantees.
+In a similar way, one may wish to introduce externally authenticated and 
+pre-shared data to the early secret derivation. This can be used to bind external 
+protocols to the TLS protocol.
```
Comment (jhoyla) [inline draft-jhoyla-tls-key-injection.md:68]: I'm wondering here if we should say "bind TLS to the external protocol". Bindings can have a direction (inward, outward, or both), and matching the order used in the previous sentence might stop people assuming that all bindings are from the external protocol to the TLS channel. But maybe I'm overthinking this.

File: draft-jhoyla-tls-key-injection.md [hunk 3]
```diff
@@ -83,44 +81,18 @@ extra context implied.
...
-The use of the imp binder derivations means that the PSK ID MUST have an opaque
-context field.
-For example, a PSK with additional data could be constructed as follows.
-
-~~~
-  struct {
-    opaque external_identity<1...2^16-1>;
-    opaque context<0...2^16>;
-  } PSKIDWithAdditionalData
-~~~
```
Comment (jhoyla) [inline draft-jhoyla-tls-key-injection.md:105]: We should include this struct. This should be the "minimum requirement".

File: draft-jhoyla-tls-key-injection.md [hunk 4]
```diff
@@ -83,44 +81,18 @@ extra context implied.
              |
              +-----> Derive-Secret(., "ext binder"
              |                      | "res binder"
-             |                      | "imp ext binder"
-             |                      | "imp res binder", "")
```
Comment (jhoyla) [inline draft-jhoyla-tls-key-injection.md:87]: To make this general we need to include the "imp res binder". In particular it's necessary if we want to use this structure for ESNI. ESNI can be used with either an external PSK or a resumption PSK, and we need to be able to distinguish those cases.

**Review by jhoyla — CHANGES_REQUESTED** [review]

File: draft-jhoyla-tls-key-injection.md (hunk 3 above)
Comment (chris-wood) [reply draft-jhoyla-tls-key-injection.md:105]: Hmm... I'm not sure. I agree that there's some minimum requirement, but I think we can state that without giving a structure.

**Pull request merged**